### ISSUE TITLE
[QUERY] Azure Identity package is generating a lot of errors during local debugging

### ISSUE THREAD

**Issue by pavoldecky** [Azure/azure-sdk-for-net#44928]
### Library name and version

Azure.Identity 1.12.0

### Query/Question

Hello 

We have Asp.Net core application that is using multiple Azure resources (KeyVault, App configuration, App Insights, Service bus, Blob Storage). Previously we used connection strings and secrets to connect to all Azure Resources.

We have a requirement, to start using managed identity to connect to Azure resources. After the switch, we have a problem during local development and debugging. Below you can see the startup code, which is set up logging for Azure identity, ClientSecretCredential, and all Azure resources.

```
WebApplicationBuilder builder = WebApplication.CreateBuilder(args);

using Stream stream = new FileStream(@"logs.txt", FileMode.OpenOrCreate, FileAccess.Write);

using StreamWriter streamWriter = new StreamWriter(stream)
{
    AutoFlush = true
};

using AzureEventSourceListener customListener = new AzureEventSourceListener((args, message) =>
{
    streamWriter.WriteLine(message);
    streamWriter.WriteLine("---");
}, EventLevel.LogAlways);

ClientSecretCredentialOptions options = new()
{
    IsUnsafeSupportLoggingEnabled = true,
    Diagnostics =
    {
        IsAccountIdentifierLoggingEnabled = true,
        IsLoggingContentEnabled = true,
        IsLoggingEnabled = true,
        IsDistributedTracingEnabled = true,
        IsTelemetryEnabled = true
    },
    Retry = { 
        NetworkTimeout = TimeSpan.FromSeconds(60)
    }
};

TokenCredential tokenCredential = new ClientSecretCredential(
    builder.Configuration[TENANT_ID],
    builder.Configuration[CLIENT_ID],
    builder.Configuration[CLIENT_SECRET],
    options
    );

builder.Services.AddAzureClients(azure =>
{
    azure.UseCredential(tokenCredential);

    azure.AddSecretClient(new Uri(kvUri));

    azure.AddServiceBusClientWithNamespace(serviceBusNamespace)
        .WithName("name1");

    azure.AddServiceBusAdministrationClientWithNamespace(serviceBusNamespace)
        .WithName("name2");

    azure.AddBlobServiceClient(new Uri(storageEnpoint));
});

builder.Configuration.AddAzureAppConfiguration(options =>
{
    options.Connect(new Uri(appConfigurationUrl), tokenCredential)
           .Select(KeyFilter.Any, ConfigKeys.ALL_SERVICES_LABEL);
});

builder.Services.Configure<TelemetryConfiguration>(config =>
{
    config.SetAzureTokenCredential(tokenCredential);
});

builder.Services.AddApplicationInsightsTelemetry();
```

First of all, it looks like the logic is working as expected, so our application is working. Unfortunately, during the debug session in Visual Studio, we are receiving plenty of errors in our logs and it looks like, all of them are from the Azure identity package OR MSAL package used by Azure Identity. 

Here are 2 errors, together with all additional logs with correlation Identity

**1.**
```
CorrelationId = 78e6aed6-ce5a-4b87-89e6-7de062bf2bee

False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] MSAL MSAL.NetCore with assembly version '4.61.3.0'. CorrelationId(78e6aed6-ce5a-4b87-89e6-7de062bf2bee)
----
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] === AcquireTokenForClientParameters ===
SendX5C: False
ForceRefresh: False
-----
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] 
=== Request Data ===
Authority Provided? - True
Scopes - https://management.azure.com//.default
Extra Query Params Keys (space separated) -
ApiId - AcquireTokenForClient
IsConfidentialClient - True
SendX5C - False
LoginHint ? False
IsBrokerConfigured - False
HomeAccountId - False
CorrelationId - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee
UserAssertion set: False
LongRunningOboCacheKey set: False
Region configured:
---
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] === Token Acquisition (ClientCredentialRequest) started:
	 Scopes: https://management.azure.com//.default
	Authority Host: login.microsoftonline.com
---
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] [Region discovery] Not using a regional authority. 
---
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] [Instance Discovery] Skipping Instance discovery because it is disabled. 
---
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:19Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] [ClientCredentialRequest] Acquiring a token from the token provider.
---
False MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - 78e6aed6-ce5a-4b87-89e6-7de062bf2bee] Exception type: Azure.Identity.CredentialUnavailableException
---> Inner Exception Details
Exception type: System.Threading.Tasks.TaskCanceledException
---> Inner Exception Details
Exception type: System.Threading.Tasks.TaskCanceledException
---> Inner Exception Details
Exception type: System.Threading.Tasks.TaskCanceledException
To see full exception details, enable PII Logging. See https://aka.ms/msal-net-logging
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.HttpConnectionWaiter`1.WaitForConnectionWithTelemetryAsync(HttpRequestMessage request, HttpConnectionPool pool, Boolean async, CancellationToken requestCancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)

=== End of inner exception stack trace ===
To see full exception details, enable PII Logging. See https://aka.ms/msal-net-logging
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Azure.Core.Pipeline.HttpClientTransport.ProcessSyncOrAsync(HttpMessage message, Boolean async)
   at Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)

=== End of inner exception stack trace ===
To see full exception details, enable PII Logging. See https://aka.ms/msal-net-logging
   at Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Identity.ManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ImdsManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)

=== End of inner exception stack trace ===
To see full exception details, enable PII Logging. See https://aka.ms/msal-net-logging
   at Azure.Identity.ImdsManagedIdentitySource.AuthenticateAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityClient.AuthenticateCoreAsync(Boolean async, TokenRequestContext context, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityClient.AppTokenProviderImpl(AppTokenProviderParameters parameters)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.SendTokenRequestToAppTokenProviderAsync(ILoggerAdapter logger, CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.GetAccessTokenAsync(CancellationToken cancellationToken, ILoggerAdapter logger)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.<>c__DisplayClass11_1.<<RunAsync>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.Identity.Client.Utils.StopwatchService.MeasureCodeBlockAsync(Func`1 codeBlock)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
```

**2.**
```

CorrelationId = d0130522-ad03-47d3-ac7a-17c786552d46

True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] MSAL MSAL.NetCore with assembly version '4.61.3.0'. CorrelationId(d0130522-ad03-47d3-ac7a-17c786552d46)
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] === AcquireTokenForClientParameters ===
SendX5C: False
ForceRefresh: False
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] 
=== Request Data ===
Authority Provided? - True
Client Id - ****
Scopes - https://monitor.azure.com//.default
Redirect Uri - https://replyUrlNotSet
Extra Query Params Keys (space separated) - 
ClaimsAndClientCapabilities - 
Authority - https://login.microsoftonline.com/****/
ApiId - AcquireTokenForClient
IsConfidentialClient - True
SendX5C - False
LoginHint - 
IsBrokerConfigured - False
HomeAccountId - 
CorrelationId - d0130522-ad03-47d3-ac7a-17c786552d46
UserAssertion set: False
LongRunningOboCacheKey set: False
Region configured: 
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] === Token Acquisition (ClientCredentialRequest) started:
	Authority: https://login.microsoftonline.com/****/
	Scope: https://monitor.azure.com//.default
	ClientId: ****
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] [Region discovery] Not using a regional authority. 
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] Access token is not expired. Returning the found cache entry. [Current time (07/09/2024 10:19:20) - Expiration Time (07/09/2024 11:12:46 +00:00) - Extended Expiration Time (07/09/2024 11:12:46 +00:00)]
Request [575875e1-114c-4754-b4ba-687862c24a13] exception System.Threading.Tasks.TaskCanceledException: The operation was cancelled because it exceeded the configured timeout of 0:00:01. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.HttpConnectionWaiter`1.WaitForConnectionWithTelemetryAsync(HttpRequestMessage request, HttpConnectionPool pool, Boolean async, CancellationToken requestCancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Azure.Core.Pipeline.HttpClientTransport.ProcessSyncOrAsync(HttpMessage message, Boolean async)
   at Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   --- End of inner exception stack trace ---
   at Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46] 
	=== Token Acquisition finished successfully:
---
True MSAL 4.61.3.0 MSAL.NetCore .NET 8.0.6 Microsoft Windows 10.0.22621 [2024-07-09 10:19:20Z - d0130522-ad03-47d3-ac7a-17c786552d46]  AT expiration time: 9. 7. 2024 11:12:46 +00:00, scopes: https://monitor.azure.com//.default. source: Cache
```
So I cannot find the root cause for both of the errors. As we are executing multiple calls to Azure, I am not able to see which Azure calls can cause exceptions. Is it possible that both errors are from Azure Identity directly and are not caused by any additional Azure call?

For the first one, I tried to enable PII Logging. According to source code, MSAL enable PII logging flag should be set by IsUnsafeSupportLoggingEnabled according to this [changelog ](https://github.com/Azure/azure-sdk-for-net/blob/97050dc0f16ee1b156784667bf968a06608fb816/sdk/identity/Azure.Identity/CHANGELOG.md?plain=1#L133)
But although I set this property to true, the result is still the same. Do you know what I am doing wrong? Maybe if I can see PII details, the error message can give me more info. 

For the second issue, I also tried to extend network timeout by setting NetworkTimeout in the Retry object in TokenCredentialOptions, but again, nothing changed and the error message is still the same. 

So can anyone please help me to better understand these errors and to check what can be done and how to avoid them? 

As I mentioned, looks like the app logic is working, so one solution how to get rid of those errors will be to set the minimal log level to Fatal for Azure.Identity, but I would like to understand what is the problem. 

Thank you very much for any help

### Environment

Windows 11, .NET 8.0.300, Visual Studio 17.10

**Comment by christothes** [comment]
Hi @pavoldecky - 
Are you using Microsoft.Identity.Web in your application? If so, I believe this comes from that using our DefaultAzureCredential internally. If not, are you using DefaultAzureCredential in your app directly?

Regardless, these errors are expected and harmless. They are an artifact of the DefaultAzureCredential environment detection logic, which tries to determine if a managed identity is available before moving on to other credentials. This behavior is explained [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/README.md#defaultazurecredential).

**Comment by pavoldecky** [comment]
Hi @christothes 
Thank you very much for your hint and explanation. 

I wanted to avoid using DefaultAzureCredential, so I am using ClientSecretCredential for local debugging and WorkloadIdentityCredential in Release mode inside AKS. But there was still one place in the code, which I used DefaultAzureCredential. If I changed it to ClientSecretCredential all errors disappear. 

But the funny thing here is that the errors with DefaultAzureCredential are shown only if I used it with ArmClient from Azure.ResourceManager package, If I use DefaultAzureCredential in other Azure packages, no errors are shown. So I will check what I am doing wrong. 

Thank you very much for your help. This issue be closed.
